[BUG] Azure Open AI produces incorrect json for tools without parameters

Please provide as much details as possible, this will help us to deliver a fix as soon as possible.
Thank you!

**Describe the bug**
The Azure OpenAI lib doesn't produce correct json for functions/tools without parameters. Current implementation (in InternalAzureOpenAiHelper) sets the parameter field to `{}` but it should be `{
"type": "object", 
"properties": {}
}`

**Log and Stack trace**
`Caused by: com.azure.core.exception.HttpResponseException: Status code 400, 
"{
  "error": {
    "message": "'{}' is not of type 'object' - 'functions.2.parameters'",
    "type": "invalid_request_error",
    "param": null,
    "code": null
  }
}`

**To Reproduce**
Use a tool without parameters

**Expected behavior**
No exception should be thrown.

**Please complete the following information:**
- LangChain4j version: e.g. 0.25.0
- Java version: e.g. 21
- Spring Boot version (if applicable): e.g. 3.2.1

**Additional context**
functions soon to be replace by tools in the api but the schema for the parameters property seem to be the same so I guess it still makes sense to fix it.

--- a/langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/InternalAzureOpenAiHelper.java
+++ b/langchain4j-azure-open-ai/src/main/java/dev/langchain4j/model/azure/InternalAzureOpenAiHelper.java
@@ -16,7 +16,6 @@ import com.azure.core.http.policy.HttpLogOptions;
 import com.azure.core.http.policy.RetryOptions;
 import com.azure.core.util.BinaryData;
 import com.azure.core.util.HttpClientOptions;
-import com.azure.identity.DefaultAzureCredentialBuilder;
 import dev.langchain4j.agent.tool.ToolExecutionRequest;
 import dev.langchain4j.agent.tool.ToolParameters;
 import dev.langchain4j.agent.tool.ToolSpecification;
@@ -167,10 +166,15 @@ class InternalAzureOpenAiHelper {
         return functionDefinition;
     }
 
+    private static final Map<String, Object> NO_PARAMETER_DATA = new HashMap<>();
+    static {
+        NO_PARAMETER_DATA.put("type", "object");
+        NO_PARAMETER_DATA.put("properties", new HashMap<>());
+    }
     private static BinaryData toOpenAiParameters(ToolParameters toolParameters) {
         Parameters parameters = new Parameters();
         if (toolParameters == null) {
-            return BinaryData.fromString("{}");
+            return BinaryData.fromObject(NO_PARAMETER_DATA);
         }
         parameters.setProperties(toolParameters.properties());
         parameters.setRequired(toolParameters.required());

--- a/langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiChatModelIT.java
+++ b/langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiChatModelIT.java
@@ -123,7 +123,7 @@ public class AzureOpenAiChatModelIT {
         Response<AiMessage> response = model.generate(Collections.singletonList(userMessage), toolSpecification);
 
         AiMessage aiMessage = response.content();
-        assertThat(aiMessage.text()).isBlank();
+        assertThat(aiMessage.text()).isNull();
 
         assertThat(aiMessage.toolExecutionRequests()).hasSize(1);
         ToolExecutionRequest toolExecutionRequest = aiMessage.toolExecutionRequests().get(0);
@@ -189,7 +189,7 @@ public class AzureOpenAiChatModelIT {
         Response<AiMessage> response = model.generate(Collections.singletonList(userMessage), noArgToolSpec);
 
         AiMessage aiMessage = response.content();
-        assertThat(aiMessage.text()).isBlank();
+        assertThat(aiMessage.text()).isNull();
 
         assertThat(aiMessage.toolExecutionRequests()).hasSize(1);
         ToolExecutionRequest toolExecutionRequest = aiMessage.toolExecutionRequests().get(0);

--- a/langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiChatModelIT.java
+++ b/langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiChatModelIT.java
@@ -114,13 +114,13 @@ public class AzureOpenAiChatModelIT {
         // This test will use the function called "getCurrentWeather" which is defined below.
         String toolName = "getCurrentWeather";
 
-        ToolSpecification toolSpecification = ToolSpecification.builder()
+        ToolSpecification weatherToolSpec = ToolSpecification.builder()
                 .name(toolName)
                 .description("Get the current weather")
                 .parameters(getToolParameters())
                 .build();
 
-        Response<AiMessage> response = model.generate(Collections.singletonList(userMessage), toolSpecification);
+        Response<AiMessage> response = model.generate(Collections.singletonList(userMessage), weatherToolSpec);
 
         AiMessage aiMessage = response.content();
         assertThat(aiMessage.text()).isBlank();
@@ -159,6 +159,42 @@ public class AzureOpenAiChatModelIT {
         assertThat(response2.content().text()).isNotBlank();
         assertThat(response2.content().text()).contains("t-shirt");
         assertThat(response2.finishReason()).isEqualTo(STOP);
+    }
+
+    @ParameterizedTest(name = "Deployment name {0} using {1}")
+    @CsvSource({
+            "gpt-35-turbo, gpt-3.5-turbo",
+            "gpt-4,        gpt-4"
+    })
+    void should_call_function_with_no_argument(String deploymentName, String gptVersion) {
+        ChatLanguageModel model = AzureOpenAiChatModel.builder()
+                .endpoint(System.getenv("AZURE_OPENAI_ENDPOINT"))
+                .serviceVersion(System.getenv("AZURE_OPENAI_SERVICE_VERSION"))
+                .apiKey(System.getenv("AZURE_OPENAI_KEY"))
+                .deploymentName(deploymentName)
+                .tokenizer(new OpenAiTokenizer(gptVersion))
+                .logRequestsAndResponses(true)
+                .build();
+
+        UserMessage userMessage = userMessage("What time is it?");
+
+        // This test will use the function called "getCurrentDateAndTime" which takes no arguments
+        String toolName = "getCurrentDateAndTime";
+
+        ToolSpecification noArgToolSpec = ToolSpecification.builder()
+                .name("getCurrentDateAndTime")
+                .description("Get the current date and time")
+                .build();
+
+        Response<AiMessage> response = model.generate(Collections.singletonList(userMessage), noArgToolSpec);
+
+        AiMessage aiMessage = response.content();
+        assertThat(aiMessage.text()).isBlank();
+
+        assertThat(aiMessage.toolExecutionRequests()).hasSize(1);
+        ToolExecutionRequest toolExecutionRequest = aiMessage.toolExecutionRequests().get(0);
+        assertThat(toolExecutionRequest.name()).isEqualTo(toolName);
+        assertThat(toolExecutionRequest.arguments()).isEqualTo("{}");
     }
 
     private static ToolParameters getToolParameters() {

--- a/langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiChatModelIT.java
+++ b/langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiChatModelIT.java
@@ -182,7 +182,7 @@ public class AzureOpenAiChatModelIT {
         String toolName = "getCurrentDateAndTime";
 
         ToolSpecification noArgToolSpec = ToolSpecification.builder()
-                .name("getCurrentDateAndTime")
+                .name(toolName)
                 .description("Get the current date and time")
                 .build();
 

--- a/langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiChatModelIT.java
+++ b/langchain4j-azure-open-ai/src/test/java/dev/langchain4j/model/azure/AzureOpenAiChatModelIT.java
@@ -114,13 +114,13 @@ public class AzureOpenAiChatModelIT {
         // This test will use the function called "getCurrentWeather" which is defined below.
         String toolName = "getCurrentWeather";
 
-        ToolSpecification weatherToolSpec = ToolSpecification.builder()
+        ToolSpecification toolSpecification = ToolSpecification.builder()
                 .name(toolName)
                 .description("Get the current weather")
                 .parameters(getToolParameters())
                 .build();
 
-        Response<AiMessage> response = model.generate(Collections.singletonList(userMessage), weatherToolSpec);
+        Response<AiMessage> response = model.generate(Collections.singletonList(userMessage), toolSpecification);
 
         AiMessage aiMessage = response.content();
         assertThat(aiMessage.text()).isBlank();

